Refactor, extend benchmarks and incorporate into CI

glwagner opened the pull request:
This PR refactors the `/benchmark` directory and incorporates the benchmark scripts into CI. It removes `test/benchmark_tests.jl`.

Our ultimate objective is to run a curated set of benchmarks regularly, and develop a graphic that tracks the evolution of the benchmarks over time and as performance optimization work is incorporated. This will ensure that the benchmark scripts stay up to date with current API. The benchmark scripts should also be runnable standalone for manual benchmarking. Eventually, we would also like to upload profile artifacts for inspection.

For now, I have moved the existing scripts in `benchmarks` into an "archive" folder. I think we should delete these, since they will be superceded. However, one question is how/whether we should also incorporate the code in `benchmarks/src`. I think this code is useful and nice for graphical display, but I also think there is a benefit to having simple benchmarks that consists as single scripts. @ali-ramadhan perhaps you can weigh in here since you developed that code originally.

cc @simone-silvestri @giordano

glwagner committed:
reorganize benchmarks

glwagner committed:
rm benchmark tests

glwagner committed:
move env to archive

glwagner committed:
fixes

glwagner committed:
fix warning

glwagner committed:
use ab2 by default

ali-ramadhan commented:
Love this! Most of the code in https://github.com/CliMA/Oceananigans.jl/blob/main/benchmark/src/Benchmarks.jl is to help keep the benchmark scripts short so it'll probably still be useful if we plan on having multiple types of benchmarks.

Some things to think about:
1. There are a lot of benchmarks in https://github.com/CliMA/Oceananigans.jl/tree/main/benchmark Which are still worth benchmarking and which are worth benchmarking regularly? 
2. To track the evolution of the benchmarks, will benchmarks be stored somewhere to be compared with future benchmarks?
3. How clean and reproducible do we want the benchmarks to be? Ideally benchmarks will be run on the same machine with the same GPU with no other processes utilizing the CPU significantly or consuming significant amounts of memory. Benchmarking on CI servers like Nautilus may not produce the cleanest benchmarks but maybe if we include the CPU and memory utilization for context then the benchmarks can be interpreted accordingly.

glwagner commented:
> 1. There are a lot of benchmarks in https://github.com/CliMA/Oceananigans.jl/tree/main/benchmark Which are still worth benchmarking and which are worth benchmarking regularly?

I am hoping we can have a system that allows us to benchmark all important situations with just a few scripts which can be run in CI to ensure they do not go stale. Do you think the existing benchmarks (the ones we care about) will incorporate into a new framework?

glwagner committed:
Embellishments coded up on Fri call

simone-silvestri commented on benchmark/benchmark_hi_res_hydrostatic.jl:
```diff
@@ -2,8 +2,8 @@ using Oceananigans
-    for _ in 1:10
+function many_steps!(model, Nt; Δt=1e-3)
+    for _ in 1:Nt
         time_step!(model, Δt)
```
I would use NVTX here to make sure the complete time step shows in the nvidia profiler report
```suggestion
function many_steps!(model, Nt; Δt=1e-3)
    for _ in 1:Nt
        NVTX.@range "time step" begin
            time_step!(model, Δt)
        end
```

glwagner committed:
Update benchmark/benchmark_hi_res_hydrostatic.jl

Co-authored-by: Simone Silvestri <silvestri.simone0@gmail.com>

glwagner committed:
Merge branch 'main' into refactor-benchmarks

simone-silvestri committed:
Merge branch 'main' into refactor-benchmarks

glwagner committed:
update benchmark

glwagner committed:
fix

glwagner committed:
add environment for the benchmarks

glwagner committed:
Merge remote-tracking branch 'origin/main' into refactor-benchmarks

Commit:
Fix channel configuration

Commit:
Add ClimaOcean bathymetry to Tripolar grid as default
If bathy is set to false, will revert to gaussian islands

glwagner committed:
Fix indentation for bathymetry assignment

Commit:
Pin Thermodynamics.jl to be v0.15.2
Thermodynamics.jl introduces changes in v0.15.3/4 which are incompatible with ClimaOcean v0.8.10

giordano commented on benchmark/Project.toml:
```diff
@@ -1,25 +1,13 @@
-PrettyTables = "08abe8d2-0d0c-5749-adfa-8a2ac140af0d"
-PyPlot = "d330b81b-6aea-500a-939a-2ce795aea3ee"
 SeawaterPolynomials = "d496a93d-167e-4197-9f49-d3af4ff8fe40"
+Thermodynamics = "b60c26fb-14c3-4610-9d3e-2d17fe7ff00c"
```
Pinning Thermodynamics shouldn't be necessary anymore: https://github.com/JuliaRegistries/General/pull/146294 prevents incompatible versions from being installed with ClimaOcean

glwagner committed:
Update benchmark_hi_res_hydrostatic.jl
